File: LeetCode/竞赛/周赛/251-300/周赛-292/6057. 统计值等于子树平均值的节点数.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int averageOfSubtree(TreeNode root) {
        int [] res = dfs(root);
        return res[0];
    }

    private int [] dfs(TreeNode node) {
        if (node == null) return new int [] {0, 0, 0};
        int [] left = dfs(node.left);
        int [] right = dfs(node.right);

        int res = left[0] + right[0];
        int sum = left[1] + right[1] + node.val;
        int cnt = left[2] + right[2] + 1;
        if (node.val == sum / cnt) res++;
        return new int[] {res, sum, cnt};
    }

    
}

File: LeetCode/竞赛/周赛/251-300/周赛-292/6057. 统计值等于子树平均值的节点数.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def averageOfSubtree(self, root: Optional[TreeNode]) -> int:
        res = 0

        def dfs(root):
            nonlocal res
            if not root: return 0, 0
            lSsum, lCnt = dfs(root.left)
            rSsum, rCnt = dfs(root.right)
            ssum = root.val + lSsum + rSsum
            cnt = 1 + lCnt + rCnt
            if (root.val == ssum // cnt): res += 1
            return ssum, cnt
        dfs(root)
        return res


